torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -483,55 +486,68 @@ def __init__(self, config):
             self.shared_experts = MLP(
                 config=config, intermediate_size=intermediate_size
             )
-        # Two shuffle method supported:
-        # 1. "torch_all_to_all"
-        # 2. "symm_mem" (see `setup_symm_mem` below)
-        self.shuffle_method = "torch_all_to_all"
 
-    # This function is used to create a symm mem buffer for MoE. It is for
+    # This function is used to create a symm mem buffer for MoE's. It is for
     # shuffling tokens fully "on-device", as compared to traditional torch
     # all_to_all APIs which requrie a GPU-to-CPU sync of the splits.  If a user
     # calls this function, the `shuffle_method` would switch from
     # `torch_all_to_all` to `symm_mem`.
-
-    # Status: supports inference. For training, this is disabled for now. Reason
-    # is that autograd requires tensors not be modified a second time, this
-    # conflicts with our wish of sharing the symm mem across layers and/or
-    # PP microbatches.
-    def setup_symm_mem(self, dtype, device):
+    def setup_symm_mem(
+        self, dtype: torch.dtype, device: torch.device, microbatches: int
+    ):
         # Switch shuffle method
         self.shuffle_method = "symm_mem"
+        self.microbatches = microbatches
+        self.curr_send = 0
+        self.curr_gather = 0
+
+        # Assuming worst case, 2x tokens are routed to one EP rank
+        overflow = 2
+        OnDeviceAllToAllV.max_output_len = (
+            self.config.max_seq_len * self.num_experts_per_tok * overflow
+        )
+
         # Symmetric memory buffers are shared by all MoE instances across
-        # layers, so we only need to initialize them once
-        if MoE.token_send_buf is not None:
+        # layers, we only need to initialize them once
+        if len(MoE.token_send_buf):
             return
 
+        # But they are not shared across microbatches
         # Input buffer for DP-to-EP shuffle
-        MoE.token_send_buf = symm_mem.empty(
-            self.config.max_seq_len
-            * self.num_experts_per_tok,  # seq len * top k (flattened)
-            self.config.hidden_size,  # hidden dim
-            dtype=dtype,
-            device=device,
-        )
-        # Number of tokens to send to EP peers, aka. input splits
-        MoE.input_splits = symm_mem.empty(
-            self.ep_size, dtype=torch.int64, device=device
-        )
-        # Number of tokens to receive from EP peers, aka. output splits
-        MoE.output_splits = symm_mem.empty(
-            self.ep_size, dtype=torch.int64, device=device
-        )
+        for _ in range(self.microbatches):
```

Comment:
we start to bleed more parallelism info into the model def..  seems logical to do it this way for now but conceptually what other options would we have? 

normally the pipeline runtime manages the lifetime of the microbatch activations indirectly:  it holds onto microbatch activations until running backward, then frees them via cuda caching allocator for reuse.

I guess for symm memm, we could rely on an allocator too?  we could just ask the symm_memm allocator for new buffers on each forward pass. when would they be freed- would they naturally be kept alive as 'activations' and freed automatically (returned to the cache) once the pipelining runtime cleared their backward pass?  I think that would be a lot cleaner than having to keep track of current microbatch inside model state.

Reply:
> we start to bleed more parallelism info into the model def..

True and false. I sort of treat `setup_symm_mem` as a "side" helper to model definition in strict term. This is also the practice when people need to setup KV cache. 

> we could rely on an allocator too? 

Yep, great idea.